src/substrait/function_registry.py

```diff
@@ -0,0 +1,340 @@
+from substrait.gen.proto.parameterized_types_pb2 import ParameterizedType
+from substrait.gen.proto.type_pb2 import Type
+from importlib.resources import files as importlib_files
+import itertools
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import Any, Optional, Union
+from .derivation_expression import evaluate
+
+import yaml
+import re
+
+
+DEFAULT_URI_PREFIX = "https://github.com/substrait-io/substrait/blob/main/extensions"
+
+
+# mapping from argument types to shortened signature names: https://substrait.io/extensions/#function-signature-compound-names
+_normalized_key_names = {
+    "i8": "i8",
+    "i16": "i16",
+    "i32": "i32",
+    "i64": "i64",
+    "fp32": "fp32",
+    "fp64": "fp64",
+    "string": "str",
+    "binary": "vbin",
+    "boolean": "bool",
+    "timestamp": "ts",
+    "timestamp_tz": "tstz",
+    "date": "date",
+    "time": "time",
+    "interval_year": "iyear",
+    "interval_day": "iday",
+    "interval_compound": "icompound",
+    "uuid": "uuid",
+    "fixedchar": "fchar",
+    "varchar": "vchar",
+    "fixedbinary": "fbin",
+    "decimal": "dec",
+    "precision_time": "pt",
+    "precision_timestamp": "pts",
+    "precision_timestamp_tz": "ptstz",
+    "struct": "struct",
+    "list": "list",
+    "map": "map",
+}
+
+
+def normalize_substrait_type_names(typ: str) -> str:
+    # First strip nullability marker
+    typ = typ.strip("?").lower()
```

Comment:
The nullability marker doesn't necessarily go at the end.  Will this work as expected for `struct?<str,str>`?

Reply:
Here are a set of type test cases that might be useful:   https://github.com/substrait-io/substrait-cpp/blob/main/src/substrait/type/tests/TypeTest.cpp

Reply:
thanks, I was convinced it was the other way around. I'll have to handle that in `to_parameterized_type` method as well, then.

Reply:
Made most of the changes, also added proper handling of scalar function's `nullability` field. Some stuff still needs a bit of a rework, but I added a bunch of TODOs for those.